sdk/identity/azure-identity/test/ut/client_assertion_credential_test.cpp

```diff
@@ -0,0 +1,406 @@
+
+std::string GetAssertion_Throw(Azure::Core::Context const&)
+{
+  throw std::runtime_error(
+      "The test is not expected to call this function used for assertion callback.");
+}
+
+std::string GetAssertion_Test(Azure::Core::Context const&) { return "sample-assertion"; }
+
+TEST(ClientAssertionCredential, GetCredentialName)
+{
+  std::string tenantId = "01234567-89ab-cdef-fedc-ba8976543210";
+  std::string clientId = "fedcba98-7654-3210-0123-456789abcdef";
+  std::string serviceConnectionId = "abc";
+  std::string systemAccessToken = "123";
+
+  ClientAssertionCredential const cred(tenantId, clientId, GetAssertion_Throw);
+
+  EXPECT_EQ(cred.GetCredentialName(), "ClientAssertionCredential");
+}
+
+TEST(ClientAssertionCredential, GetOptionsFromEnvironment)
+{
+  std::string tenantId = "01234567-89ab-cdef-fedc-ba8976543210";
+  std::string clientId = "fedcba98-7654-3210-0123-456789abcdef";
+
+  {
+    std::map<std::string, std::string> envVars = {{"AZURE_AUTHORITY_HOST", ""}};
+    CredentialTestHelper::EnvironmentOverride const env(envVars);
+
+    ClientAssertionCredentialOptions options;
+    ClientAssertionCredential const cred(tenantId, clientId, GetAssertion_Throw, options);
+    EXPECT_EQ(cred.GetCredentialName(), "ClientAssertionCredential");
+
+    EXPECT_EQ(options.AuthorityHost, "https://login.microsoftonline.com/");
+  }
+
+  {
+    std::map<std::string, std::string> envVars = {{"AZURE_AUTHORITY_HOST", "foo"}};
+    CredentialTestHelper::EnvironmentOverride const env(envVars);
+
+    ClientAssertionCredentialOptions options;
+    options.AuthorityHost = "bar";
+    EXPECT_EQ(options.AuthorityHost, "bar");
+  }
+
+  {
+    std::map<std::string, std::string> envVars
+        = {{"AZURE_AUTHORITY_HOST", "https://microsoft.com/"}};
+    CredentialTestHelper::EnvironmentOverride const env(envVars);
+
+    ClientAssertionCredentialOptions options;
+    EXPECT_EQ(options.AuthorityHost, "https://microsoft.com/");
+  }
+}
+
+TEST(ClientAssertionCredential, InvalidArgs)
+{
+  std::string tenantId = "01234567-89ab-cdef-fedc-ba8976543210";
+  std::string clientId = "fedcba98-7654-3210-0123-456789abcdef";
+
+  // Empty Tenant ID
+  {
+    TokenRequestContext trc;
+    trc.Scopes.push_back("https://storage.azure.com/.default");
+
+    ClientAssertionCredential const cred("", clientId, GetAssertion_Throw);
+    EXPECT_THROW(cred.GetToken(trc, {}), AuthenticationException);
+  }
+
+  // Invalid Tenant ID
+  {
+    TokenRequestContext trc;
+    trc.Scopes.push_back("https://storage.azure.com/.default");
+
+    ClientAssertionCredential const cred("!=invalidTenantId=!", clientId, GetAssertion_Throw);
+    EXPECT_THROW(cred.GetToken(trc, {}), AuthenticationException);
+  }
+
+  // Empty client ID
+  {
+    TokenRequestContext trc;
+    trc.Scopes.push_back("https://storage.azure.com/.default");
+
+    ClientAssertionCredential const cred(tenantId, "", GetAssertion_Throw);
+    EXPECT_THROW(cred.GetToken(trc, {}), AuthenticationException);
+  }
```

Comment:
So, if clientId is empty, are you sure it is an error? It's is not the case for other credentials (other than AzurePipelines, but I am not sure about that one too)

Reply:
Ah, my memory was wrong about clientId being optional, deleting my comment.